worlds/landstalker/__init__.py

```diff
@@ -0,0 +1,245 @@
+            for entrance in created_entrances:
+                entrance.access_rule = make_path_requirement_lambda(self.player, [], [self.regions_table['tibor']])
+
+    def set_rules(self):
+        Rules.create_rules(self.multiworld, self.player, self.regions_table, self.dark_region_ids)
+
+        # In "Reach Kazalt" goal, player doesn't have access to Kazalt, King Nole's Labyrinth & King Nole's Palace.
+        # As a consequence, all locations inside those regions must be excluded and the teleporter from
+        # King Nole's Cave to Kazalt must go to the end region instead.
+        if self.get_setting("goal") == LandstalkerGoal.option_reach_kazalt:
+            kazalt_tp = self.multiworld.get_entrance("king_nole_cave -> kazalt", self.player)
+            kazalt_tp.connected_region = self.regions_table["end"]
+
+            EXCLUDED_REGIONS = [
+                "kazalt",
+                "king_nole_labyrinth_pre_door",
+                "king_nole_labyrinth_post_door",
+                "king_nole_labyrinth_exterior",
+                "king_nole_labyrinth_fall_from_exterior",
+                "king_nole_labyrinth_raft_entrance",
+                "king_nole_labyrinth_raft",
+                "king_nole_labyrinth_sacred_tree",
+                "king_nole_labyrinth_path_to_palace",
+                "king_nole_palace"
+            ]
+
+            for location in self.multiworld.get_locations(self.player):
+                if location.parent_region.name in EXCLUDED_REGIONS:
+                    location.progress_type = LocationProgressType.EXCLUDED
+
+    def get_starting_health(self):
+        spawn_id = self.get_setting('spawn_region')
+        if spawn_id == "destel":
+            return 20
+        elif spawn_id == "verla":
+            return 16
+        elif spawn_id in ["waterfall", "mercator", "greenmaze"]:
+            return 10
+        else:
+            return 4
+
+    def generate_output(self, output_directory: str) -> None:
+        self.adjust_shop_prices()
+
+        self.hints = Hints.generate_random_hints(self.multiworld, self.player)
+        self.hints["Lithograph"] = Hints.generate_lithograph_hint(self.multiworld, self.player, self.jewel_items)
+        self.hints["Oracle Stone"] = f"It shows {self.dark_dungeon_id}\nenshrouded in darkness."
+
+        self.can_fill_slot_data.set()
+
+    def adjust_shop_prices(self):
+        # Calculate prices for items in shops once all items have their final position
+        unknown_items_price = 250
+        earlygame_price_factor = 1.0
+        endgame_price_factor = 2.0
+        factor_diff = endgame_price_factor - earlygame_price_factor
+
+        global_price_factor = self.get_setting("shop_prices_factor") / 100.0
+
+        spheres = list(self.multiworld.get_spheres())
+        sphere_id = 0
+        sphere_count = len(spheres)
+        for sphere in spheres:
+            for location in sphere:
+                if location.player != self.player or location.type_string != 'shop':
```

Comment:
should this be and? you're calling this method per player

Reply:
alternatively, use stage_post_fill to do this for all landstalker players at once. `get_spheres()` isn't cached, so calling it multiple times is very expensive.

Reply:
the caching part is addressed with https://github.com/ArchipelagoMW/Archipelago/pull/2410, but i still think just looping through it a single time for all landstalker worlds should be more performant, if possible.